Grandcentral Routing (temporary while DHT is shown to be correct)

briantigerchow opened the pull request:
@jbenet @mappum @whyrusleeping 

TODOs

- [x] pick a serialization format for provider records. perhaps just proto message
- [x] ~~extract database `Backend` from `grandcentral.Server`~~ create datastore
- [x] pass config value (server | client) in from UX
- [x] pick a peer/address/host for remote routing server
- [x] finally, make this test pass: https://build.protocol-dev.com/job/ipfs-network-test-add-cat/

jbenet commented:
@briantigerchow is this still ongoing?  dht seems to be better off now, but i'd like to see some large (thousands of nodes) tests before really relying on it.

briantigerchow commented:
@jbenet This is ready for some initial testing

briantigerchow commented:
@jbenet @whyrusleeping

GCR is ready for some tire kicking.

I wrote an integration test to see simulate 2 clients bootstrapping to 8 servers. The test performs a 100 MB file transfer between the two clients. GCR functions under these conditions.

Next, I deployed GCR-enabled bootstrap servers to canarynet and wrote a script to add and cat files from clients in an automated, controlled manner.

It works as follows
```Go
// adder
for i := 0; i < math.MaxInt32; i++ {
  r := generateRandomBytes(i, seed) // i bytes of data given by seed
  coreunix.Add(node, r)
    fmt.Println(added file |i| with seed |seed|)
  time.Sleep(time.Second)
}
```

```Go
// catter
for i := 0; i < math.MaxInt32; i++ {
  k := determineHashOfFile(i, seed)
  coreunix.Cat(n, k)
  fmt.Println(sucessfully catted |i| bytes with seed |seed|)
  time.Sleep(time.Second)
}
```

The test chokes after just a few files/bytes of data. Through it's simplicity, the test scenario has revealed several bugs in other areas of the codebase. The investigation is ongoing.

whyrusleeping commented:
Does GCR interop with the DHT at all?

briantigerchow commented:
> Does GCR interop with the DHT at all?

They both conform to the routing interface, so they can be combined using a `routing.MultiRouter`. Clients can use both GCR and DHT. Specific semantics haven't been agreed upon.

Although GCR does not currently expose its proto, it uses the same proto as the DHT, so messages are cross-compatible.

cc @jbenet

briantigerchow commented:
@jbenet @whyrusleeping 

This is ready for CR. 

The work of merging GCR into the main `ipfs` binary is underway. I have chosen to introduce these changes in a separate PR. This is to allow this PR to focus on the core routing operations, and the subsequent one on integration concerns.

briantigerchow commented:
ugh @ vendoring 7KLOC aws packages

jbenet commented on cmd/ipfs-gateway-fs/main.go:
```diff
@@ -53,13 +65,19 @@ func run() error {
+	if err != nil {
+		return err
+	}
+	node, err := core.NewIPFSNode(ctx, core.OnlineWithRouting(repo, corerouting.GrandCentralServer(enhanced)))
```
does this mean all gateways will use `S3` to back up everything? it would be nice to be able to control this: only pin certain things we specifically want. Otherwise anyone could abuse this by having our gateways download TB of stuff.

How it should work:
- ipfs gateways are a cache and GC
- a subset (or a different set) pin some objects we chose to seed indefinitely

briantigerchow replied:
The datastore passed to GCR is used _only_ for routing records. The node still has a local datastore.

added comment daaee56a78eadb8da742455ff1ab384dca89d46b

jbenet replied:
Ah, okay. may be worth adding a note here too. given this is the `ipfs-gateway-fs` which we may direct other users to setup.

It may be worth actually splitting this into a different binary, as you had it earlier.

briantigerchow replied:
Yeah. The routing servers should probably be in a separate binary altogether.

jbenet commented:
One major point: the name `grand central` is great as an internal name. it is not good for external users, as (a) they will be trying to understand the meaning of the underlying algorithm from the name, (b) it is misleading. It is not fully centralized (users can delegate to a different set of servers), and it is not our goal to keep this in the long run. Can we switch the code itself from using "GrandCentral" in its naming to `DelegatedRouting` or something that evokes a similar meaning?

jbenet commented on core/corerouting/core.go:
```diff
@@ -0,0 +1,85 @@
+			if err := node.PeerHost.Connect(ctx, info); err != nil {
+				return nil, err // TODO
+			}
+		}
```
handle the TODO? this will cause connections before the network is ready to listen.

briantigerchow replied:
How do you want to integrate the bootstrap method into the core? The existing bootstrap procedure appears to preclude polymorphism.

briantigerchow replied:
@jbenet Requesting advice about moving the `Connect` calls into `Bootstrap()`.

https://github.com/jbenet/go-ipfs/pull/454#discussion_r23925859

jbenet replied:
We should actually remove the `dht.Connect` from there and make it all routing-agnostic. 

the dht should be taking all live connections into its routing table now (by notification), and we could always do so in the periodic DHT Bootstrap.

briantigerchow replied:
@jbenet I took a good, hard look at the core Bootstrap routines. Based on your experience with the implementation, I think you'll want to step in and make the necessary changes.

jbenet commented on core/corerouting/core.go:
```diff
@@ -0,0 +1,85 @@
+		}
+
+		// TODO right now, I think this has a hidden dependency on the
+		// bootstrap peers provided to the core.Node. Careful...
```
could you expand on this? where would that come to bear?

briantigerchow replied:
removed. no longer a concern now that gcr servers are defined separately from the bootstrap peers.

8f345fdbf668b9e652cf7fc10653c96d6b45861d

jbenet commented on routing/grandcentral/proxy/standard.go:
```diff
@@ -46,28 +74,51 @@ func (px *standard) SendMessage(ctx context.Context, m *dhtpb.Message) error {
 func (px *standard) SendRequest(ctx context.Context, m *dhtpb.Message) (*dhtpb.Message, error) {
-	if err := px.Host.Connect(ctx, peer.PeerInfo{ID: px.Remote}); err != nil {
+	var err error
+	for _, i := range rand.Perm(len(px.Remotes)) {
```
maybe add a comment describing the process?
```go
// try each remote sequentially, use the first success.
```

briantigerchow replied:
85e978701efa7608c404d349d086b77cb1bbe62e

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -42,91 +42,53 @@ func (s *Server) HandleRequest(ctx context.Context, p peer.ID, req *dhtpb.Messag
-	//  FIXME threw everything into this switch statement to get things going.
-	//  Once each operation is well-defined, extract pluggable backend so any
-	//  database may be used.
+	log.EventBegin(ctx, "routingMessageReceived", req, p, s.local).Done() // TODO may need to differentiate between local and remote
```
`defer`

briantigerchow replied:
f2c6cdcac28d5949e028a95c3eb1d0b30a7b2cc4

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -42,91 +42,53 @@ func (s *Server) HandleRequest(ctx context.Context, p peer.ID, req *dhtpb.Messag
 			return "", nil
 		}
+		response.Record = rawRecord
 		// TODO before merging: if we know any providers for the requested value, return those.
```
Todo?

briantigerchow replied:
I didn't realize I'd left behind so many breadcrumbs. I'll re-read, address outstanding self-TODOs and ping for CR afterward.

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -42,91 +42,53 @@ func (s *Server) HandleRequest(ctx context.Context, p peer.ID, req *dhtpb.Messag
 		return p, response
 
 	case dhtpb.Message_PUT_VALUE:
 		// TODO before merging: verifyRecord(req.GetRecord())
```
todo?

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -42,91 +42,53 @@ func (s *Server) HandleRequest(ctx context.Context, p peer.ID, req *dhtpb.Messag
-			return "", nil
-		}
+		putRoutingRecord(s.routingBackend, util.Key(req.GetKey()), req.GetRecord())
 		return p, req // TODO before merging: verify that we should return record
```
todo?

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -42,91 +42,53 @@ func (s *Server) HandleRequest(ctx context.Context, p peer.ID, req *dhtpb.Messag
+				// Connectedness: TODO
+			},
+		}
+		response.CloserPeers = dhtpb.PeerRoutingInfosToPBPeers(pri)
```
This means that the router's internal cache is the only way to get peer addresses. This may be fine, given that getting addresses isnt a problem atm, but it certainly should be described in some high level description of how GCR works. I would've expected addresses to also be stored in the datastore.

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -42,91 +42,53 @@ func (s *Server) HandleRequest(ctx context.Context, p peer.ID, req *dhtpb.Messag
+		// FIXME(btc): do we want to store these locally? I think the
+		// storeProvidersToPeerstore behavior is straight from the DHT message
+		// handler.
+		storeProvidersToPeerstore(s.peerstore, p, req.GetProviderPeers())
```
fixme? we should be storing the provider records in the datastore.

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -138,3 +100,92 @@ func (s *Server) handleMessage(
+	}
+	for _, provider := range newRecords {
+		mergedRecords[provider.GetId()] = provider // overwrite old record if new exists
+	}
```
important to make it clear in the implementation design doc (maybe in the package's godoc or above this method) that currently there is a last-writer-wins race between GCR nodes.

This should also timeout/expire records. We must not store provider records indefinitely. the current expiration is something like 12/24hrs.

jbenet commented on routing/grandcentral/server.go:
```diff
@@ -138,3 +100,92 @@ func (s *Server) handleMessage(
+	for _, provider := range providers {
+		providerID := peer.ID(provider.GetId())
+		if providerID != p {
+			log.Errorf("provider message came from third-party %s", p)
```
`log.Event` ?

jbenet replied:
there should be this check as well in `putRoutingProviders` or this should return an error. Otherwise the record _will_ be stored (despite us not collecting their address).

briantigerchow commented:
addressed comments

- [x] reset bootstrap executable to its state on master (routingd is separate executable)
- [x] address TODOs
- [ ] put connection setup in Bootstrap method
- [ ] enable record validation. (complains the remote peer's public key is not present in peerstore)

briantigerchow commented:
@jbenet what's the status? `routingd` initialization successful now that bootstrapping is in place?

jbenet commented:
@briantigerchow i haven't tested `routingd` -- how?

briantigerchow commented:
I'll deploy a few servers and give it a shot.

briantigerchow commented:
@jbenet for some reason, clients are unable to dial servers during routing Bootstrap

```
12:49:40.210 CRITI       core: {<peer.ID QmaCpD> [/ip4/104.131.131.82/tcp/4001]} <autogenerated>:20
12:49:40.400 CRITI supernode/: dial attempt failed <autogenerated>:20
12:49:40.400 CRITI supernode/: {<peer.ID QmZDYP> [/ip4/107.170.215.87/tcp/4001]} <autogenerated>:20
12:49:40.606 CRITI supernode/: dial attempt failed <autogenerated>:20
12:49:40.606 CRITI supernode/: unable to bootstrap to any supernode routers <autogenerated>:20
```

Nothing fancy happening.

briantigerchow commented:
`ipfs_routingd` is deployed on solarnet (port 4002)

use `ipfs daemon --routing=supernode` to connect

briantigerchow commented:
I'm unsure why, when querying the peerstore for the public key of the remote peer, it isn't present in the host/network peerstore. (The public key is required for signature verification). Thoughts?

cc @jbenet @whyrusleeping

jbenet commented:
> I'm unsure why, when querying the peerstore for the public key of the remote peer, it isn't present in the host/network peerstore. (The public key is required for signature verification). Thoughts?

Can you be more specific? who is this host? nodes only have public keys of hosts (a) they've directly  connected + done the secio handshake with, (b) they've found public keys through the routing system. we dont persist keys yet (perhaps we should be).

briantigerchow commented:
@jbenet 
```Go
// creates and signs a record for the given key/value pair
func makeRecord(ps peer.Peerstore, p peer.ID, k u.Key, v []byte) (*pb.Record, error) {
	blob := bytes.Join([][]byte{[]byte(k), v, []byte(p)}, []byte{})
	sig, err := ps.PrivKey(p).Sign(blob)
	if err != nil {
		return nil, err
	}
	return &pb.Record{
		Key:       proto.String(string(k)),
		Value:     v,
		Author:    proto.String(string(p)),
		Signature: sig,
	}, nil
}
```

```Go
func verify(ps peer.Peerstore, r *dhtpb.Record) error {
	v := make(record.Validator)
	v["pk"] = record.ValidatePublicKeyRecord
	p := peer.ID(r.GetAuthor())
	pk := ps.PubKey(p)
	if pk == nil {
		return fmt.Errorf("do not have public key for %s", p) // <----
	}
	if err := v.VerifyRecord(r, pk); err != nil {
		return err
	}
	return nil
}
```

jbenet commented:
@briantigerchow i understand where it's happening. 

> Can you be more specific? who is this host?

briantigerchow commented:
The client is a routing client and the server is a routing server. 

(They're connected to each other. This implies they've performed a secio handshake.)

jbenet commented:
`mocknet` clients dont perform secio handshake. IIRC they currently dont put keys in peerstore. this can change.

briantigerchow commented:
Ah. Even if it doesn't actually do the encrypted handshake, the side-effects should be the same as the real thing.

jbenet commented:
> Even if it doesn't actually do the encrypted handshake, the side-effects should be the same as the real thing.

Well it's not about secio vs mocknet. getting the keys from the secure conn is strictly an optimization. it wont happen with raw (non encrypted) connections either. Also, it doesnt work for records from people you're not directly connected to. A provider record is currently not signed by the provider in the dht, but it will be in the future, so having a record without having the key is normal.

What this boils down to is that keys need to be retrieved if not already there. this means:
- if not directly connected, query the routing system for either (a) addresses to the peer (findpeer), or (b) the key itself (get/put value).
- it directly connected, ask _it_ for its key (this can happen in `identify`. it presently doesnt send the pk but it can.)

briantigerchow commented:
@jbenet @whyrusleeping 

* `ipfs_routingd` supports redis as a backend (see flags in `cmd/ipfs_routingd/main.go`)
* supernode client has learned to prioritize servers based on XOR distance, comparing remote.ID vs msg.Key (previously uniformly randomized)
* supernode client has learned to replicate PUTVALUE, ADDPROVIDER messages to multiple servers (default: 2)

briantigerchow commented:
@jbenet RFCR

jbenet commented on cmd/ipfs_routingd/.go-ipfs/config:
```diff
@@ -0,0 +1,60 @@
+{
```
is this config supposed to be here? 

With the privkey here, anyone can pretend to be this client. (not that it matters atm.)

jbenet commented on cmd/ipfs_routingd/.go-ipfs/config:
```diff
@@ -0,0 +1,60 @@
+  },
+  "Datastore": {
+    "Type": "leveldb",
+    "Path": "/Users/btc/go/src/github.com/jbenet/go-ipfs/cmd/ipfs_routingd/.go-ipfs/datastore"
```
if this is supposed to be here, the path shouldnt be absolute.

jbenet commented on cmd/ipfs_routingd/.go-ipfs/config:
```diff
@@ -0,0 +1,60 @@
+    "MaxSizeMB": 250,
+    "MaxBackups": 1,
+    "MaxAgeDays": 0
+  }
```
if the config file is here just for the routing server addrs, maybe those can be added in a sh script, or hardcoded into the snr code.

jbenet commented on cmd/ipfs_routingd/.go-ipfs/datastore/LOG:
```diff
@@ -0,0 +1,9 @@
+=============== Feb 6, 2015 (MST) ===============
```
ok it looks like the `.go-ipfs` dir was added by mistake (all these leveldb files)

jbenet commented on cmd/ipfs_routingd/main.go:
```diff
@@ -0,0 +1,151 @@
+
+func enhanceDatastore(d datastore.Datastore) (datastore.ThreadSafeDatastore, error) {
+	// TODO cache
+	return ds2.CloserWrap(syncds.MutexWrap(d)), nil
```
why does the s3 datastore need to be mutex-wrapped?

briantigerchow replied:
doesn't need to be. good catch. declared the s3ds as a ThreadSafeDatastore and dropped the mutex.

jbenet commented on routing/supernode/client.go:
```diff
@@ -79,13 +81,23 @@ func (c *Client) GetValue(ctx context.Context, k u.Key) ([]byte, error) {
-	// TODO wrap this to hide the dialer and the local/remote peers
-	msg.ProviderPeers = pb.PeerInfosToPBPeers(c.dialer, []peer.PeerInfo{peer.PeerInfo{ID: c.local}}) // FIXME how is connectedness defined for the local node
-	return c.proxy.SendMessage(ctx, msg)                                                             // TODO wrap to hide remote
+	// FIXME how is connectedness defined for the local node
```
> how is connectedness defined for the local node

connectedness is still whether the node is directly connected to another, at the time of request.

briantigerchow replied:
what is the connectedness value for self?

jbenet commented on test/supernode_client/main.go:
```diff
@@ -0,0 +1,256 @@
+	"/ip4/178.62.8.190/tcp/4001/ipfs/QmdXzZ25cyzSF99csCQmmPZ1NTbWTe8qtKFaZKpZQPdTFB",
+	"/ip4/188.166.8.195/tcp/4001/ipfs/QmNU1Vpryj5hfSmybSYHnS497ttgy9aNJ3T2B8wY2uMso4",
+	"/ip4/188.226.225.73/tcp/4001/ipfs/QmYHGLxLfHFd9veUt159LMhLD8LqhDjBUS9ZehLkPap6od",
+	"/ip4/192.241.209.121/tcp/4001/ipfs/QmNPM851fxWb3qjLV66a8Nqtw7fBnMw2UrdFQidSdBQnvS",
```
should these be pulled from config?

briantigerchow replied:
yeah. fixed

jbenet commented:
a few comments, otherwise lgtm

briantigerchow commented:
@jbenet removed the stray `.go-ipfs` dir and rebased/erased it out of the git history. declared the s3 datastore as thread safe.

jbenet commented:
@briantigerchow changes lgtm, thank you.

jbenet commented:
Everything LGTM! let me know how you want to proceed with this merge. My constraints are:

- trivial to manage
- easy for users to know how to switch between dht/snr (this could be just an example or a note in the readme we can point them to)
- we should not enable it by default until we're certain it is better than dht. (i.e. run it alongside for a while and we can play with it?). We could become certain of this after a few hours of trying it out

briantigerchow commented:
Feel free to merge. The servers are already running. Users Pass flag "--routing=supernode" to turn it on.

whyrusleeping commented:
We're getting that fault panic on tip a lot.

briantigerchow commented:
Fault panic on tip?

whyrusleeping commented:
@briantigerchow  Yeah, on the jenkins "golang tip" build, we keep seeing this really weird fault, that we arent sure the cause of
